### ISSUE TITLE
[Urgent - WithIn One Week][SF][UI5-List][A11y][NVDA] Screen Reader

### ISSUE THREAD

**Issue by sfsf-xweb-sh** [SAP/ui5-webcomponents#3806]
**Dear colleagues,**
I am from SuccessFactors Web Experience Shanghai team.
we use NVDA to test a11y issues in our project

**Reproduce Step:**
1. Open [https://github.wdf.sap.corp/pages/xweb/assignment/app.html?mock#/assignment-detail/ao/2](https://github.wdf.sap.corp/pages/xweb/assignment/app.html?mock#/assignment-detail/ao/2)
2. Click the applicants tab

**Bug Description**

**Issue 1. The Applicants is a `ui5-li-groupheader`, it should not be announced as a list item.**
Expected Behavior: The `ui5-li-groupheader` should not be announced as a list item
![assignment-ui5-list-a11y-1](https://user-images.githubusercontent.com/78520316/131809675-91043527-a166-4fa9-8912-e218b6240cbc.png)

**Issue 2. The Applicants is a `ui5-li-groupheader`, So the total count of list item is 6 ( `ui5-li-groupheader` should not be counted into total number ) , but screen reader reads number 7 ( actually `ui5-li-groupheader` is counted into total number  )**
![assignment-ui5-list-a11y-3](https://user-images.githubusercontent.com/78520316/131813689-bec3319e-0510-4534-89c5-1845757be7a1.png)

**Issue 3. Selected state of radio button(checked/ not checked) should be announced by screen reader.**
In this issue, Applied item is a selected item, but the selected state is not announced by screen reader .
Expected behavior: the selected state should be announced by screen reader .
![assignment-ui5-list-a11y-2](https://user-images.githubusercontent.com/78520316/131810645-46fc33c3-9b1e-4c3e-9a52-47859dc15c50.png)

**Issue 4. Listbox missing accessible name**
When the listbox role is added to an element, screen reader should  announce the label and role of the listbox when it gets focus.
![assignment-ui5-list-a11y-4](https://user-images.githubusercontent.com/78520316/131816362-f252a027-6a3a-40d0-8715-d3a295eb134b.png)

**Issue 5: Role of radio buttons should be announced for screen reader users**
Role of radio buttons should be announced for screen reader users
Ensure radio buttons have accessible name using aria-label or aria-labelledby attribute
![assignment-ui5-list-a11y-5](https://user-images.githubusercontent.com/78520316/131817059-b5bc38f8-d3c1-4e3b-b526-2617b368e79e.png)

**Issue 6: If I set accessible-name attribute to ui5-list, the text will be announced twice by screen reader**
In this case, the text All is announced twice by screen reader
![assignment-ui5-list-a11y-6](https://user-images.githubusercontent.com/78520316/131818328-7ba13f23-c47a-4172-882f-796d887a77a3.png)

**Comment by dobrinyonkov** [comment]
Hello @sfsf-xweb-sh, 

I have looked at your reports. 

1, 2 - The group header is implemented similarly to the one used in sapui5 and it is by design. I raised a discussion internally to consider your feedback.

3 - Seems to be reproducible on Chrome and Firefox with plain HTML. It is the same with NVDA and JAWS, I found some issues reported in their repositories and raised a discussion internally to find if it is the expected behavior.
_Note: Not selected(false) state is announced. And selected(true) state is announced upon selection only._

JAWS (open) - https://github.com/FreedomScientific/VFO-standards-support/issues/440
NVDA (closed but still reproducible) - https://github.com/nvaccess/nvda/issues/6562


4 - The accessible name of the `ui5-list` is up to the application to be set. An API `accessible-name`  is exposed on the `ui5-list` which value we forward to the listbox's `aria-label` attribute in the shadow dom.

5 - We will add accessible name to the radio button.

6 - The twice announcement in this case comes from the fact that the list item textContent is the same as the `accessible-name` value - "All".  The `accessible-name` should be used to additionally label the item, setting the same value as the item text would always lead to the same twice announcement.

Kind Regards,
Dobrin

**Comment by sfsf-xweb-sh** [comment]
Hi @dobrinyonkov 

Thanks for your quick reply.
looking forward to your internal discussion

**Comment by fifoosid** [comment]
Hi @sfsf-xweb-sh 

Do you need the fixes for this issue in 0.29 or 0.31 releases?

**Comment by dobrinyonkov** [comment]
Hi @sfsf-xweb-sh,

we have merged 3 pull requests addressing the points: 1, 2, 3, 5.
For the other issues to be resolved please see my first comment above.

Kind Regards,
Dobrin

**Comment by sfsf-xweb-sh** [comment]
Hi @fifoosid  @dobrinyonkov 

Thanks for your great job.
I need this fixed to be released in version 0.31.x ( sf version )
The current sf version is 0.31.16. Maybe this fix can be released with version 0.31.17

Thanks 

Kind Regards,
Hubery

**Comment by fifoosid** [comment]
Hi @sfsf-xweb-sh 

This will be released in the next 0.31.17 version.

**Comment by sfsf-xweb-sh** [comment]
Hi @dobrinyonkov @fifoosid 

I have tried issue 5 in 0.31.17, but this issue is still available though `accessible-name` is indeed added for `ui5-radiobutton`

> Issue 5: Role of radio buttons should be announced for screen reader users
![](https://user-images.githubusercontent.com/78520316/131817059-b5bc38f8-d3c1-4e3b-b526-2617b368e79e.png)

<img width="1098" alt="Screen Shot 2021-09-15 at 11 09 30 AM" src="https://user-images.githubusercontent.com/78520316/133378052-657b14f4-d53d-4fee-be05-95988ff44d0d.png">

Please reopen this ticket

Thanks

**Comment by fifoosid** [comment]
Hi @sfsf-xweb-sh 

I tested it out with the provided link and everything seems to be working as expected. The only issue of this kind is not related to the ui5-li:
![image](https://user-images.githubusercontent.com/16850339/133398356-b4e3f9de-7cbc-4330-808e-96142ce7ca42.png)

If you inspect the DOM element, you can that now it has the ```aria-label``` attribute:
![image](https://user-images.githubusercontent.com/16850339/133398503-5e1f63ea-3191-437a-8f6b-15871a92496e.png)

You might be looking at a cached version

**Comment by sfsf-xweb-sh** [comment]
Hi @fifoosid 
Thanks for your reply.

I know the fact that the Dom has `aria-label` attribute now.
But the fact is the role of radio button ( radio ) is not announced by screenreader. and the `aria-label: Item Selection` can not be announced by screen reader too

**Comment by dobrinyonkov** [comment]
Hi @sfsf-xweb-sh,

By design, those radio buttons are not meant to be interactive and they are not reachable via keyboard, therefore they are not announced. 
They have tabindex attribute of "-1".

The reproted issue of AXE is that every toggle field should have an accessible name.

Kind Regards,
Dobrin

**Comment by sfsf-xweb-sh** [comment]
Hi @dobrinyonkov  @fifoosid 

I mean that, the screen reader ( NVDA ) will read "Offer Rejected 2 not selected 5 of 6" now.
But the screen reader should read "Option 5 radio button Offer Rejected 2 not checked 5 of 6"

Because I use ui5-list ( mode: SingleSelectBegin ), so it's behavior should be similar to radio button.
If screen reader don't announce "Option 5 radio button" in this case, user  will be confused about what this option will do

<img width="1098" alt="Screen Shot 2021-09-15 at 11 09 30 AM" src="https://user-images.githubusercontent.com/78520316/133609890-d48cce7e-9e57-41d0-9baa-97653cf1ec02.png">

**Comment by dobrinyonkov** [comment]
Hi @sfsf-xweb-sh,

Thank you for clarifying!
I will route your feedback to the specification owners for further discussion.

Kind Regards,
Dobrin

**Comment by sfsf-xweb-sh** [comment]
Hi @dobrinyonkov 

Is there any update about my latest comment ?
Thanks

Kind Regards,
Hubery

**Comment by dobrinyonkov** [comment]
Hi @sfsf-xweb-sh,

we are still working on your feedback.
I will update the issue once we have clear statement.

Thanks and Best Regards,
Dobrin

**Comment by sfsf-xweb-sh** [comment]
Hi @dobrinyonkov 

Is there any update ?

Kind Regards,
Hubery

**Comment by dobrinyonkov** [comment]
Hi @sfsf-xweb-sh,

Those checkboxes and radio buttons in this case are visual indication about the selection mode being applied to the list. It should be clear that these are not real controls and there is no need to announce them as such. These are solely there to “transport” the info that this is a single or multiselectable list. However, there is no such announcment to the screen reader users so we will add a label "Single Select" or "Multi Select" to the list to differentiate the two modes.

Kind Regards,
Dobrin

**Comment by dobrinyonkov** [comment]
Hi @sfsf-xweb-sh,

the above is now added to the `ui5-list` implementation. On entry a tutor message about the type of selection of the list is announced.

Kind Regards,
Dobrin

**Comment by ilhan007** [comment]
Hello @sfsf-xweb-sh we released 0.31.20, including one more fix to the List - on entry a tutor message about the type of selection of the list is announced.
